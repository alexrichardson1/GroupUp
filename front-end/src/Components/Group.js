const Group = ({ group }) => {
  const { leader, teammates, requirements, adRequirements } = group;

  const temp = () => console.log("hello world!");

  let keyGen1 = 0;
  let keyGen2 = 0;

  return (
    <div onClick={temp} className="group">
      <h1>{leader}'s Group</h1>
      <ul>
        {/* <li>
          Other Members:
          <ul>
            {teammates.map((name) => (
              <li key={keyGen1++}>{name}</li>
            ))}
          </ul>
        </li> */}
        <li>
          Requirements:
          <ul>
            {Object.keys(requirements).map((key) => (
              <li key={keyGen2++}>{key + ": " + requirements[key]}</li>
            ))}
            {/* {for (let key in requirements.keys()) {
              if(requirements.hasOwnProperty()) {
                <li key={keyGen2++}>key</li>
              }
            }
            } */}

            {/* {requirements.map((req) => (
              <li key={keyGen2++}>{req}</li>
            ))} */}
          </ul>
        </li>
        <li>Additional Requirements: {adRequirements}</li>
      </ul>
      <button className="btn" onClick={null}>
        Join Group
      </button>
    </div>
  );
};

export default Group;
