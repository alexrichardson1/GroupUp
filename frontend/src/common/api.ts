import { ActiveT, GroupT, ProjectT, UserT } from "types/types";
import axios from "axios";
import { config } from "Constants";

const dummyGroup: GroupT = {
  id: -1,
  leader: "LEADER",
  maxmembers: -1,
  teammates: ["ALICE", "BOB"],
  requirements: ["REQUIREMENT"],
  adrequirements: "ADDITIONAL REQUIREMENTS",
  projectid: -1,
  leaderemail: "DUMMY@EMAIL.COM",
  posted: "TIME",
};

const dummyProject: ProjectT = {
  id: -1,
  name: "NAME",
  requirements: ["REQUIREMENT"],
  description: "DESCRIPTION",
  hours: 0,
  date: "DATE",
  location: "LOCATION",
};

const dummyUser: UserT = {
  id: -1,
  fullname: "FULL NAME",
  email: "DUMMY@EMAIL.COM",
  password: "PASSWORD",
  activefilter: ["FILTER"],
  lastlogin: "LAST LOGIN",
  groupsid: -1,
};

const dummyActive: ActiveT = {
  id: -1,
  fullname: "FULL NAME",
  email: "DUMMY@EMAIL.COM",
  filters: [],
};

async function addGroup(data: GroupT) {
  await axios
    .post(`${config.API_URL}/group/add`, data)
    .then((res) => {
      console.log(`Group added: ${res.data}`);
    })
    .catch((error) => {
      console.error(error);
    });
}

async function joinGroup(groupId: number, name: string) {
  await axios
    .post(`${config.API_URL}/group/join`, {
      groupid: groupId,
      name: name,
    })
    .then((res) => {
      console.log(`Group joined: ${res.data}`);
      alert("Successfully joined group!");
    })
    .catch((error) => {
      console.error(error);
    });
}

async function getGroup(groupId: number) {
  const group = dummyGroup;
  await axios
    .post(`${config.API_URL}/group/one`, {
      groupid: groupId,
    })
    .then((res) => {
      const group = res.data;
      console.log(group);
    })
    .catch((error) => {
      console.error(error);
    });
  return group;
}

async function getGroups() {
  var result: GroupT[] = [dummyGroup];
  await axios
    .get(`${config.API_URL}/group`)
    .then((res) => {
      result = res.data;
    })
    .catch((error) => {
      console.error(error);
    });
  return result;
}

async function addProject(data: ProjectT) {
  await axios
    .post(`${config.API_URL}/project/add`, data)
    .then((res) => {
      console.log(`Project added: ${res.data}`);
    })
    .catch((error) => {
      console.error(error);
    });
}

async function getProject(id: number) {
  var result = dummyProject;
  await axios
    .post(`${config.API_URL}/project/one`, {
      projectid: id,
    })
    .then((res) => {
      result = res.data;
    })
    .catch((error) => {
      console.error(error);
    });
  return result;
}

async function getProjects() {
  var result: ProjectT[] = [dummyProject];
  await axios
    .get(`${config.API_URL}/project`)
    .then((res) => {
      result = res.data;
    })
    .catch((error) => {
      console.error(error);
    });
  return result;
}

async function getUser(email: string) {
  var result = dummyUser;
  await axios
    .post(`${config.API_URL}/user/one`, {
      email: email,
    })
    .then((res) => {
      result = res.data;
    })
    .catch((error) => {
      console.error(error);
    });
  return result;
}

async function getUsers() {
  var result: UserT[] = [dummyUser];
  await axios
    .get(`${config.API_URL}/user`)
    .then((res) => {
      result = res.data;
    })
    .catch((error) => {
      console.error(error);
    });
  return result;
}

async function getActive() {
  var result = dummyActive;
  await axios
    .get(`${config.API_URL}/active/one`)
    .then((res) => {
      result = res.data;
    })
    .catch((error) => {
      console.error(error);
    });
  return result;
}

async function setActive(email: string, fullName: string) {
  await axios
    .post(`${config.API_URL}/active/email/update`, {
      email: email,
      fullname: fullName,
    })
    .then((res) => {
      console.log(`Updated active: ${res.data}`);
    })
    .catch((error) => {
      console.error(error);
    });
}

async function setFilters(userEmail: string, filters: any[]) {
  await axios
    .post(`${config.API_URL}/user/activefilter/update`, {
      filters: filters,
      email: userEmail,
    })
    .then((res) => {
      console.log(`Added filters for: ${userEmail}`);
      console.log(`Filters: ${filters}`);
      console.log(`Updated active: ${res.data}`);
    })
    .catch((error) => {
      console.error(error);
    });
}

export {
  addProject,
  getProject,
  getProjects,
  dummyProject,
  addGroup,
  joinGroup,
  getGroup,
  getGroups,
  dummyGroup,
  getUser,
  getUsers,
  dummyUser,
  getActive,
  setActive,
  setFilters,
  dummyActive,
};
