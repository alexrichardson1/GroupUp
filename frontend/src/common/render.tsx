import { Alert } from "react-bootstrap";

function allFieldsNotFilledIn(invalid: boolean) {
  if (invalid) {
    return <Alert variant="danger">Fill in all fields.</Alert>;
  }
  return null;
}

export { allFieldsNotFilledIn };
