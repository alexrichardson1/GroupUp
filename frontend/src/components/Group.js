import { Button, Card, Table } from "react-bootstrap";
import { LinkContainer } from "react-router-bootstrap";

const Group = ({ group, requirementNames }) => {
  const { id, leader, requirements, posted } = group;

  let groupDetailsLink = "/group/" + id;

  const date = new Date();

  const getPostedTimeAgo = () => {
    if (posted == null) {
      return "Date unknown";
    }
    const dayAsDate = new Date(posted.substring(0, 10));

    const hoursToMills = 3600000 * parseInt(posted.substring(11, 13));
    const minutesToMills = 60000 * parseInt(posted.substring(14, 16));
    const secondsToMills = 1000 * parseInt(posted.substring(17, 19));

    const datePostedAsNum =
      dayAsDate + hoursToMills + minutesToMills + secondsToMills;

    const difference = new Date() - datePostedAsNum;

    if (difference < 2629800000) {
      if (difference > 604800000) {
        return "Posted " + Math.floor(difference / 604800000) + " weeks ago";
      } else if (difference > 86400000) {
        return "Posted " + Math.floor(difference / 86400000) + " days ago";
      } else if (difference > 3600000) {
        return "Posted " + Math.floor(difference / 3600000) + " hours ago";
      } else {
        return "Posted less than an hour ago";
      }
    } else {
      return "Posted more than a month ago";
    }
  };

  const dbDateToNumber = (curDate) => {};

  return (
    <div>
      <Card>
        <Card.Body>
          <Card.Title>{leader}'s Group</Card.Title>
          <Card.Text>
            {/* <ListGroup.Item variant="dark">Requirements:</ListGroup.Item> */}
            <Table striped bordered hover size="sm">
              <thead>
                <tr>
                  <th>Requirement</th>
                  <th>Needed</th>
                </tr>
              </thead>
              <tbody>
                {Object.entries(requirements).map(([key, val]) => (
                  <tr>
                    <td>
                      {requirementNames[key] === "Timezone"
                        ? "Timezone (UTC +/-)"
                        : requirementNames[key]}
                    </td>
                    <td>
                      {requirementNames[key] === "Timezone"
                        ? val > -1
                          ? "+" + val
                          : val
                        : val}
                    </td>
                  </tr>
                ))}
              </tbody>
            </Table>
          </Card.Text>
          <Card.Footer className="text-muted datePosted">
            {getPostedTimeAgo()}
          </Card.Footer>
          <LinkContainer to={groupDetailsLink}>
            <Button>Find out more</Button>
          </LinkContainer>
        </Card.Body>
      </Card>
    </div>
  );
};

export default Group;
