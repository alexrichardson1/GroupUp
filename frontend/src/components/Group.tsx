import { Button, Card, Table } from "react-bootstrap";
import { LinkContainer } from "react-router-bootstrap";
import { GroupT } from "types/types";

const MONTH_IN_MS = 2629800000;
const WEEK_IN_MS = 604800000;
const DAY_IN_MS = 86400000;

interface Props {
  group: GroupT;
  requirementNames: string[];
  projectId: number;
}

const Group = ({ group, requirementNames, projectId }: Props) => {
  const { id, leader, requirements, posted } = group;

  let groupDetailsLink = "/group/" + id + "/" + projectId;

  const getPostedTimeAgo = () => {
    if (posted == null) {
      return "Date unknown";
    }
    console.log(posted);
    const dayAsDate = new Date(posted.substring(0, 10)).getTime();
    console.log("Day as milli: " + dayAsDate);

    const difference = Date.now() - dayAsDate;

    if (difference < MONTH_IN_MS) {
      if (difference > WEEK_IN_MS) {
        return "Posted " + Math.floor(difference / WEEK_IN_MS) + " weeks ago";
      } else if (difference > DAY_IN_MS) {
        return (
          "Posted " +
          Math.floor(difference / DAY_IN_MS) +
          (Math.floor(difference / DAY_IN_MS) === 1 ? " day ago" : " days ago")
        );
      } else {
        return "Posted less than a day ago.";
      }
    } else {
      return "Posted more than a month ago";
    }
  };

  return (
    <div>
      <Card>
        <Card.Body>
          <Card.Title>{leader}'s Group</Card.Title>
          <Card.Text>
            <Table striped bordered hover size="sm">
              <thead>
                <tr>
                  <th>Requirement</th>
                  <th>Needed</th>
                </tr>
              </thead>
              <tbody>
                {Object.entries(requirements).map(([key, val]) => (
                  <tr>
                    <td>
                      {requirementNames[parseInt(key)] === "Timezone"
                        ? "Timezone (UTC +/-)"
                        : requirementNames[parseInt(key)]}
                    </td>
                    <td>
                      {requirementNames[parseInt(key)] === "Timezone"
                        ? parseInt(val) > -1
                          ? "+" + val
                          : val
                        : val}
                    </td>
                  </tr>
                ))}
              </tbody>
            </Table>
          </Card.Text>
          <Card.Footer className="text-muted datePosted">
            {getPostedTimeAgo()}
          </Card.Footer>
          <LinkContainer to={groupDetailsLink}>
            <Button>Find out more</Button>
          </LinkContainer>
        </Card.Body>
      </Card>
    </div>
  );
};

export default Group;
