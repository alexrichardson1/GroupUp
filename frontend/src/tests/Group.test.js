import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import Group from "components/Group";
import { dummyGroup } from "common/api";

test("renders correctly", async () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Group
          group={dummyGroup}
          requirementNames={["1", "TYPESCRIPT", "ENGLISH"]}
          projectId={0}
        />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
