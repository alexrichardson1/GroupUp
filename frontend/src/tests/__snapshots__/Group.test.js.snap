// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="card"
  >
    <div
      className="card-body"
    >
      <div
        className="card-title h5"
      >
        ALICE
        's Group
      </div>
      <p
        className="card-text"
      >
        <table
          className="table table-sm table-striped table-bordered table-hover"
        >
          <thead>
            <tr>
              <th>
                Requirement
              </th>
              <th>
                Needed
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                1
              </td>
              <td>
                1
              </td>
            </tr>
            <tr>
              <td>
                TYPESCRIPT
              </td>
              <td>
                TYPESCRIPT
              </td>
            </tr>
            <tr>
              <td>
                ENGLISH
              </td>
              <td>
                ENGLISH
              </td>
            </tr>
          </tbody>
        </table>
      </p>
      <div
        className="text-muted datePosted card-footer"
      >
        Posted more than a month ago
      </div>
      <a
        className="btn btn-primary"
        href="/group/0/0"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        Find out more
      </a>
    </div>
  </div>
</div>
`;
